Docs: add template vars/functions reference

alexfornuto opened the pull request:
WIP

Closes #11263

Adds a single consolidated reference to all template variables and functions teleport provides for RBAC.

Commit:
add template vars/functions reference

alexfornuto commented:
FYI This is not yet ready for a full review. I'm looking to @programmerq to help fill out some missing details and weigh in on placement and formatting.

alexfornuto commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -373,3 +374,163 @@ Here is an explanation of the fields used in the `where` and `filter` conditions
+
+<Details title="External Variables"
+>
+These variables provide context from an Identity Provider or other external data source.
```
@programmerq an attempt to summarize our discussion:

```suggestion
These variables provide context from an Identity Provider or other external data source. We can't predict every variable an external data source will provide, but these are common values expected from SAML and OIDC identity providers.
```

programmerq commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -373,3 +374,163 @@ Here is an explanation of the fields used in the `where` and `filter` conditions
+
+- **Type**: Variable
+
+#### `{{internal.jwt}}`
```
I believe the `{{internal.jwt}}` variable is really only valid in the context of the templating done in application access, and not in a role template.

alexfornuto replied:
Noted below.

nklaassen replied:
this is not part of a role template, it is not part of a role at all, so I don't think it should be on this page

ptgott commented:
@alexfornuto Let me know when this is ready for copy review!

Commit:
Update docs/pages/access-controls/reference.mdx

Commit:
replace details blocks with headers

With unopened details blocks the headers cannot be directly linked to, negating their usefullness as H4s

Commit:
partial updates

alexfornuto commented:
> @alexfornuto Let me know when this is ready for copy review!

Will do. Still looking for SME signoff.

nklaassen commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -373,3 +374,188 @@ Here is an explanation of the fields used in the `where` and `filter` conditions
+
+## Template variables and functions
+
+Teleport provides several template variables and functions to enable more robust access control:
```
Is this supposed to supplement the Role Templates page? https://goteleport.com/docs/access-controls/guides/role-templates/#interpolation-rules I'm a bit worried about duplication that may drift over time

alexfornuto replied:
I think that the linked section is a good introduction to the concept, and this is a reference to all available variables. I don't really see it as duplication.

nklaassen commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -373,3 +374,188 @@ Here is an explanation of the fields used in the `where` and `filter` conditions
+
+### Internal Variables and Functions
+
+#### `{{email.local(variable)}}`
+
+- **Type**: Function
+- **Description**: Extracts the local part of an email address.
+- **Example**: `{{email.local(alice@example.com)}}` -> `alice`
+
+This function is often paired with an external variable, for example:
+`{{email.local(external.username)}}` for IdPs using email addresses as the username.
+
+#### `{{regexp.match(expression, variable)}}`
+
+- **Type**: Function
+- **Description**: Returns true if the provided string matches the specified
+  pattern.
+- **Example**: `{{regexp.match("^Email: (.*)$", external.raw_email)}}` -> `true`
+  for matching input.
```
It would be nice to know exactly where each of these are relevant, I don't think it's clear here. For example, `email.local` can be used in any templated role field like `logins` (the list of all fields that support this is long), while I believe `regexp.match` is only applicable in certain fields like `allow.impersonate.where` and `allow.request.claims_to_roles[].value`. I'm happy to help find where these are all relevant

alexfornuto replied:
Yes, help would be appreciated! Would you feel like adding that content directly, or discussing with me 1:1?

nklaassen replied:
This is a section on role templates, I'd suggest we only document the role template functions here, and not the ones found in `where` and `filter` expressions, which should be part of the `Filter Fields` section immediately above and the label expressions page. Role templates are always surrounded in `{{}}`, others are not, so actually `{{regexp.match(expression, variable)}}` is not something that you would ever see, and it is already documented on the predicate language page.

The only functions available in role templates are `email.local` and `regexp.replace`

nklaassen commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -373,3 +374,188 @@ Here is an explanation of the fields used in the `where` and `filter` conditions
+- **Example**: `{{external.databases}}` -> `["db1", "db2"]`
+- **Provider**: OIDC/SAML
+
+#### `{{external["custom_attribute"]}}`
```
I think it would be nice to document that `external.trait_name` is just a shorthand for `external["trait_name"]`, and the latter is necessary when `trait_name` includes certain special characters like space or hyphen. Same is true for `internal`. Maybe we would even just switch to only documenting the `[]` syntax? I'm not sure

alexfornuto replied:
Sounds a lot like what's on L557, but could maybe use some tweaking.

nklaassen commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -373,3 +374,188 @@ Here is an explanation of the fields used in the `where` and `filter` conditions
+- **Example**: `{{external.email}}` -> `"alice@example.com"`
+- **Provider**: OIDC/SAML
+
+#### `{{external.groups}}`
+
+- **Type**: Variable
+- **Description**: Fetches the user's group membership from the SSO provider.
+- **Example**: `{{external.groups}}` -> `["group1", "group2"]`
+- **Provider**: OIDC/SAML
+
+#### `{{external.azure_identities}}`
+
+- **Type**: Variable
+- **Description**: Fetches the user's Azure identities from the SSO provider [(2)].
+- **Example**: `{{external.azure_identities}}` -> `["identity1"]`
+- **Provider**: OIDC/SAML
+
+#### `{{external.databases}}`
```
I'm not sure about documenting all of these as separate variables, it's really just a claim the comes from your IdP, it could be anything. `groups` and `email` will probably be the most common by far, many of the others would have to be manually configured in most cases I think

alexfornuto replied:
Is this concern addressed by the content at the start of the section?

> These variables provide context from an Identity Provider or other external data source. We can't predict every variable an external data source will provide, but these are common values expected from SAML and OIDC identity providers.

If not, what would you propose we change?

nklaassen replied:
I would propose that we don't document each of these individually, we would have some general info about the `external` trait syntax, and then maybe just a simple list of common trait names (i think email and groups are common, everything else they would have been a custom trait they intentionally set up).

nklaassen commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -373,3 +374,188 @@ Here is an explanation of the fields used in the `where` and `filter` conditions
+
+## Template variables and functions
+
+Teleport provides several template variables and functions to enable more robust access control:
```
nit: not sure robust is the word here, maybe flexible, dynamic, powerful. Can we link to the dedicated template variables page as an overview here? And backlink from that page to here for a complete reference

nklaassen reviewed (commented):
sorry for the delay on this, hope my comments make sense and sound reasonable, happy to chat if not

ptgott committed:
Partially respond to nklaassen feedback

- Remove the `internal.jwt` entry
- Edit the page introduction

ptgott commented:
## Remaining TODOs

- [ ] Document all and only _role template_ functions, not functions found in other contexts (e.g., `where` and `filter` expressions). ([comment](https://github.com/gravitational/teleport/pull/25431/files#r1248253930))
- [ ] Edit the guidance on `external` variables to explain how the work and not aim to document specific `external` variables ([comment](https://github.com/gravitational/teleport/pull/25431/files#r1248248712))
- [ ] Explain the difference between `external.trait` and `externa["trait"]` ([comment](https://github.com/gravitational/teleport/pull/25431/files#r1248222905))

ptgott committed:
Edit the "External variables" section

Make this more general and explain dot vs. bracket syntax

ptgott committed:
Add a note re: OIDC claim data types

ptgott committed:
Minor formatting/clarity tweaks

ptgott committed:
Add reference section for third-party principals

ptgott committed:
Start documenting vars/funcs for Access Requests

ptgott committed:
Start adding label expression vars/funcs

ptgott committed:
Add to this

ptgott commented:
I am going to close this PR and open a new one when I complete a draft of the reference. This is because the edits I'm making are pretty extensive, and internal PRs are a lot easier to work with in our CI system.
